add associate call

unconst opened the pull request:
Creates a call here which allows a coldkey to associate a hotkey to it.
This should be on cost call, meaning the cost is in TAO and comes from the coldkey hopefully via a transaction fee. 
This is needed to allow a wallet to create hotkeys associated with itself and then use those to run apps in production without having to decrypt the coldkey
for instance, if you want to query a validator with the hotkey associated with one of its delegates.

unconst committed:
add associate call

unconst committed:
add associate call

Rubberbandits commented on runtime/Cargo.toml:
```diff
@@ -42,7 +42,7 @@ sp-inherents = { version = "4.0.0-dev", default-features = false, git = "https:/
 sp-runtime = { version = "7.0.0", default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.39" }
 sp-session = { version = "4.0.0-dev", default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.39" }
-sp-std = { version = "5.0.0", default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.39" }
+sp-std = { version = "5.0.0", default-features = false, git = "https://github.com/paritytech/substrate.giet", branch = "polkadot-v0.9.39" }
```
This typo in cargo.toml seems to break everything

SaMotlagh commented on pallets/subtensor/src/lib.rs:
```diff
@@ -1105,6 +1108,19 @@ pub mod pallet {
 
+		/// Associates the hotkey to this coldkey.
+		/// TODO(const): this call should be costly.
+		#[pallet::call_index(6)]
```
- we should not have two calls with same index. we should always add calls at the end of existing calls and increment index.
- also, how are we paying for this call?

Rubberbandits replied:
We'll use an inclusion fee instead of arbitrarily increasing call weight (which will affect block production.)

S0AndS0 commented on pallets/subtensor/src/lib.rs:
```diff
@@ -1105,6 +1108,19 @@ pub mod pallet {
+		/// Associates the hotkey to this coldkey.
+		/// TODO(const): this call should be costly.
+		#[pallet::call_index(6)]
+		#[pallet::weight((Weight::from_ref_time(2_000_000_000) /// making this an expensive call?
```
Triple-slash comments (`///`) are for doc-comments, so either double-slash or (better yet) inline comments (`/* ... */`) should be used instead, eg.

```diff
-#[pallet::weight((Weight::from_ref_time(2_000_000_000) /// making this an expensive call?
+#[pallet::weight((Weight::from_ref_time(2_000_000_000) /* making this an expensive call? */
```

S0AndS0 commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +169,49 @@ impl<T: Config> Pallet<T> {
+        Self::create_account_if_non_existent( &coldkey, &hotkey);         
+
+        // --- 4. Deposit successful event.
+        log::info!("HotkeyAssociated( coldkey:{:?} hotkey:{:?}  ) ", coldkey, hotkey );
```
Missing a coma between `cold`/`hot`-`key` within output

```diff
-log::info!("HotkeyAssociated( coldkey:{:?} hotkey:{:?} ) ", coldkey, hotkey );
+log::info!("HotkeyAssociated( coldkey:{:?}, hotkey:{:?} ) ", coldkey, hotkey );
```

S0AndS0 commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +169,49 @@ impl<T: Config> Pallet<T> {
+    ) -> DispatchResult {
+        // --- 1. Check that the caller has signed the transaction. (the coldkey of the pairing)
+        let coldkey = ensure_signed( origin )?; 
+        log::info!("do_associate( coldkey:{:?} hotkey:{:?} )", coldkey, hotkey );
```
Missing a coma between cold/hot-key within output

```diff
-log::info!("do_associate( coldkey:{:?} hotkey:{:?} )", coldkey, hotkey );
+log::info!("do_associate( coldkey:{:?}, hotkey:{:?} )", coldkey, hotkey );
```

S0AndS0 commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +169,49 @@ impl<T: Config> Pallet<T> {
+        Ok(())
+    }
+    
+	/// TODO( rusty ): this will take more care, edge cases if the hotkey ends up not having a coldkey but gets referenced somewhere.
```
Mixing tabs and spaces

```diff
-	/// TODO( rusty ): this will take more care, edge cases if the hotkey ends up not having a coldkey but gets referenced somewhere.
+    /// TODO( rusty ): this will take more care, edge cases if the hotkey ends up not having a coldkey but gets referenced somewhere.
```

S0AndS0 commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +169,49 @@ impl<T: Config> Pallet<T> {
+    //     log::info!("do_associate( coldkey:{:?} hotkey:{:?} )", coldkey, hotkey );
+
+    //     // --- 2. Removes the coldkey - hotkey pairing account.
+    //     Self::remove_association( &coldkey, &hotkey);         
```
Trailing blank space

```diff
-Self::remove_association( &coldkey, &hotkey);         
+-Self::remove_association( &coldkey, &hotkey);
```

S0AndS0 commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +169,49 @@ impl<T: Config> Pallet<T> {
+    // ) -> DispatchResult {
+    //     // --- 1. Check that the caller has signed the transaction. (the coldkey of the pairing)
+    //     let coldkey = ensure_signed( origin )?; 
+    //     log::info!("do_associate( coldkey:{:?} hotkey:{:?} )", coldkey, hotkey );
```
Missing comma

```diff
-log::info!("do_associate( coldkey:{:?} hotkey:{:?} )", coldkey, hotkey );
+log::info!("do_associate( coldkey:{:?}, hotkey:{:?} )", coldkey, hotkey );
```

S0AndS0 commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +169,49 @@ impl<T: Config> Pallet<T> {
+    //     Self::remove_association( &coldkey, &hotkey);         
+
+    //     // --- 3. Deposit successful event.
+    //     log::info!("HotkeyDeAssociated( coldkey:{:?} hotkey:{:?}  ) ", coldkey, hotkey );
```
Missing comma

```diff
-log::info!("HotkeyDeAssociated( coldkey:{:?} hotkey:{:?} ) ", coldkey, hotkey );
+log::info!("HotkeyDeAssociated( coldkey:{:?}, hotkey:{:?} ) ", coldkey, hotkey );
```

S0AndS0 reviewed (commented):
Other than noted doc-comment vs inline-comment syntax, all other comments I've made _should_ be considered pedantic and likely can be safely ignored

Commit:
Force rust-analyzer to use stable toolchain when checking

Commit:
Force rust-analyzer to use stable toolchain when checking

Commit:
Set call_index for 51, fix broken comment

Commit:
Set call_index for 51, fix broken comment

Commit:
Fix broken git repo url

Commit:
Fix broken git repo url

shibshib commented on pallets/subtensor/src/lib.rs:
```diff
@@ -1105,6 +1108,21 @@ pub mod pallet {
+		#[pallet::weight((Weight::from_ref_time(2_000_000_000)
+		.saturating_add(T::DbWeight::get().reads(27))
+		.saturating_add(T::DbWeight::get().writes(22)), DispatchClass::Normal, Pays::No))]
+		pub fn associate( 
```
How were these weights calculated? Was it benchmarked on one of the validator machines?

Rubberbandits replied:
I believe the weight was arbitrarily picked by @unconst but i’ll be benchmarking both associate and de associate calls and fix the weights.

Commit:
Add docs comment for do_associate

Commit:
Add docs comment for do_associate

Commit:
Implementation for do_deassociate

Commit:
Implementation for do_deassociate

Commit:
Add event HotkeyDeAssociated and error OtherAssociation

Commit:
Add event HotkeyDeAssociated and error OtherAssociation

Commit:
Add deassociate extrinsic

Commit:
Add deassociate extrinsic

Commit:
Add associate to pre_dispatch and validate in SignedExtension

Commit:
Add associate to pre_dispatch and validate in SignedExtension

Commit:
Fix type inference in map remove calls for deassociate

Commit:
Fix type inference in map remove calls for deassociate

camfairchild commented on pallets/subtensor/src/lib.rs:
```diff
@@ -1105,6 +1108,31 @@ pub mod pallet {
+				hotkey: T::AccountId,
+		) -> DispatchResult {
+			Self::do_deassociate(origin, hotkey)
+		}
```
```suggestion
		pub fn disassociate(
				origin:OriginFor<T>,
				hotkey: T::AccountId,
		) -> DispatchResult {
			Self::do_disassociate(origin, hotkey)
		}
```

camfairchild commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +169,77 @@ impl<T: Config> Pallet<T> {
+    pub fn do_deassociate( 
+        origin: T::RuntimeOrigin,
+        hotkey: T::AccountId, 
+    ) -> DispatchResult {
```
```suggestion
    pub fn do_disassociate( 
        origin: T::RuntimeOrigin,
        hotkey: T::AccountId, 
    ) -> DispatchResult {
```

camfairchild commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +169,77 @@ impl<T: Config> Pallet<T> {
+
+        // --- 6. Deposit successful event.
+        log::info!("HotkeyDeAssociated( coldkey:{:?} hotkey:{:?}  ) ", coldkey, hotkey );
+        Self::deposit_event( Event::HotkeyDeAssociated( coldkey, hotkey ) );
```
```suggestion
        Self::deposit_event( Event::HotkeyDisassociated( coldkey, hotkey ) );
```

camfairchild commented on pallets/subtensor/src/lib.rs:
```diff
@@ -530,6 +530,8 @@ pub mod pallet {
 		WeightsSet( u16, u16 ), // ---- Event created when a caller successfully set's their weights on a subnetwork.
 		NeuronRegistered( u16, u16, T::AccountId ), // --- Event created when a new neuron account has been registered to the chain.
+		HotkeyAssociated( T::AccountId, T::AccountId ), // --- Event created when a hotkey has been associated with a coldkey.
+		HotkeyDeAssociated( T::AccountId, T::AccountId ), // --- Event created when a hotkey has been de-associated with a coldkey.
```
```suggestion
		HotkeyDisassociated( T::AccountId, T::AccountId ), // --- Event created when a hotkey has been disassociated with a coldkey.
```

camfairchild reviewed (changes requested):
Should use `disassociate` instead

Commit:
Rename deassociate -> disassociate

Commit:
Rename deassociate -> disassociate

camfairchild commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +169,77 @@ impl<T: Config> Pallet<T> {
+        log::info!("do_associate( coldkey:{:?} hotkey:{:?} )", coldkey, hotkey );
+
+        // --- 2. Check the hotkey isn't already associated with a coldkey.
+        ensure!( !Self::hotkey_account_exists( &hotkey ), Error::<T>::AlreadyRegistered );
```
Not sure this makes sense because the hotkey isn't signing the ext

Commit:
Add NotHotkeyOwner error

Commit:
Add NotHotkeyOwner error

Commit:
Verify that hotkey is owned by coldkey in disassociate

Commit:
Verify that hotkey is owned by coldkey in disassociate

Commit:
Add initial test cases for associate+disassociate

Commit:
Add initial test cases for associate+disassociate

Commit:
Add signed_coldkey argument to associate
This argument is a payload signed by the hotkey that a coldkey is asking to be associated with, verifying ownership of both hotkey and coldkey.

Commit:
Add signed_coldkey argument to associate
This argument is a payload signed by the hotkey that a coldkey is asking to be associated with, verifying ownership of both hotkey and coldkey.

Commit:
Add Signature, PublicKey types from sr25519 into runtime

Commit:
Add Signature, PublicKey types from sr25519 into runtime

Commit:
Add signature argument to associate extrinsic

Commit:
Add signature argument to associate extrinsic

Commit:
Fix trailing space

Commit:
Fix trailing space

Rubberbandits commented:
Still working on this branch, issues with PublicKey From<&[u8]> implementation

Commit:
Rearrange imports for clarity

Commit:
Rearrange imports for clarity

Commit:
Remove <T: Config>::PublicKey for core::sr25519::Public

Commit:
Remove <T: Config>::PublicKey for core::sr25519::Public

Commit:
Pass sr25519::Public directly to sig.verify

Commit:
Pass sr25519::Public directly to sig.verify

Commit:
Fix testing with empty signature

Commit:
Fix testing with empty signature

Rubberbandits commented:
Signature checking for hotkey verification is complete, need to create testing cases with legitimate signature from hotkey.

Commit:
Add InvalidHotkeyProof error

Commit:
Add InvalidHotkeyProof error

Commit:
Remove data argument from associate

Instead of letting the user supply the data they signed, we assume they signed the coldkey directly and do the same in runtime.

Commit:
Remove data argument from associate

Instead of letting the user supply the data they signed, we assume they signed the coldkey directly and do the same in runtime.

Commit:
Add signature parameter to test environment

Commit:
Add signature parameter to test environment

Commit:
Import Ss58Codec to registration tests

Commit:
Import Ss58Codec to registration tests

Commit:
Remove data argument from associate calls in unit testing

Commit:
Remove data argument from associate calls in unit testing

Commit:
Add real signature to test_hotkey_association_ok

Commit:
Add real signature to test_hotkey_association_ok

Commit:
Import Ss58Codec to registration tests

Commit:
Merge branch 'associate' of https://github.com/opentensor/subtensor into HEAD

Commit:
Remove double import

Commit:
Don't allow caller to pass arbitrary data, instead pull from origin coldkey

Commit:
Remove unused imports

Commit:
Properly implement valid signatures into association testing

Commit:
Remove unused imports

Commit:
Merge branch 'main' into associate

camfairchild commented on .vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
+	"rust-analyzer.cargo.extraEnv": {
+		"RUSTUP_TOOLCHAIN": "stable"
```
Shouldn't this be nightly?

camfairchild replied:
also, should we commit ide settings?

Rubberbandits replied:
Oops! The reason for this is that rust-analyzer only functions with rust stable it seems. Didn't mean to commit this.

S0AndS0 replied:
> @camfairchild 
> also, should we commit ide settings?

Considering the `.vscode/tasks.json` seems to have a `tasks` list that may allow for executing code, I'd say tracking that IDE configuration to be risky, and as for other IDE configuration files I've usually read others frowning on the practice... though I have been known to use Vim comments within individual files :shrug:

Rubberbandits replied:
Removed files from tracking and added the directory to gitignore

camfairchild commented on pallets/subtensor/src/registration.rs:
```diff
@@ -169,6 +172,101 @@ impl<T: Config> Pallet<T> {
+
+		// Skip extra 0th byte.
+		let hotkey_bytes: &[u8; 32] = binding[1..].as_ref().try_into().unwrap();
+		let hotkey_sr25519 = sp_core::sr25519::Public(*hotkey_bytes);
```
Do we have to make it bytes here first? Can't go straight from `AccountId`?

Rubberbandits replied:
Public has no TryFrom implementation for U256, only H256

camfairchild reviewed (approved):
Approved, with a few comments

Commit:
Remove .vscode from tracking and add to .gitignore

SaMotlagh commented on pallets/subtensor/src/lib.rs:
```diff
@@ -1726,6 +1762,12 @@ impl<T: Config + Send + Sync + TypeInfo> SignedExtension for SubtensorSignedExte
+                    priority: Self::get_priority_vanilla(),
+                    ..Default::default()
+                })
+            }
```
Do we need to explicitly set priority for this call?

Commit:
Add benchmarks, switch to MultiSignature/MultiSigner

camfairchild commented:
Closing as stale
